src/exec/execute_pipeline.c

```diff
@@ -22,53 +22,71 @@
 #define PIPE_READ 0
 #define PIPE_WRITE 1
 
-int	execute_left_node(t_ast *node, t_ctx *ctx,
-		t_pipeline_conf *conf, int *pipe_fd)
+static t_pipe_conf	*construct_pipe_conf(void)
 {
-	int	ret;
+	t_pipe_conf	*conf;
 
-	conf->fd_out = pipe_fd[PIPE_WRITE];
-	ret = execute_ast_node(node, ctx, conf);
-	close(pipe_fd[PIPE_WRITE]);
-	return (ret);
+	conf = ft_xmalloc(sizeof(t_pipe_conf));
+	conf->prev_read = -1;
+	conf->prev_write = -1;
+	conf->next_read = STDIN_FILENO;
+	conf->next_write = -1;
+	return (conf);
 }
 
-int	execute_right_node(t_ast *node, t_ctx *ctx,
-		t_pipeline_conf *conf, int *pipe_fd)
+static void	update_pipe_conf(t_pipe_conf *conf, int pipe_fd[2])
 {
-	int	ret;
+	conf->prev_read = conf->next_read;
+	conf->prev_write = conf->next_write;
+	if (pipe_fd == NULL)
+	{
+		conf->next_read = -1;
+		conf->next_write = STDOUT_FILENO;
+	}
+	else
+	{
+		conf->next_read = pipe_fd[PIPE_READ];
+		conf->next_write = pipe_fd[PIPE_WRITE];
+	}
+}
 
-	if (conf->fd_in != STDIN_FILENO)
-		close(conf->fd_in);
-	conf->fd_in = pipe_fd[PIPE_READ];
-	if (node->type != AST_PIPE)
-		conf->fd_out = STDOUT_FILENO;
-	ret = execute_ast_node(node, ctx, conf);
-	close(pipe_fd[PIPE_READ]);
+static int	free_and_return(t_pipe_conf *conf, int ret)
+{
+	free(conf);
 	return (ret);
 }
 
-int	execute_pipeline(t_ast *node, t_ctx *ctx, t_pipeline_conf *conf)
+void	close_pipe_fd(t_pipe_conf *conf)
```

Comment:
`close_pipe_fd()` is only used in `spawn_command.c`.

Reply:
Exactly